pennylane_lightning_gpu/_serialize.py

```diff
@@ -61,59 +65,69 @@ def _obs_has_kernel(obs: Observable) -> bool:
+    assert not isinstance(o, Tensor)
 
     if use_csingle:
         ctype = np.complex64
-        obs_py = ObsStructGPU_C64
+        named_obs = NamedObsGPU_C64
     else:
         ctype = np.complex128
-        obs_py = ObsStructGPU_C128
+        named_obs = NamedObsGPU_C128
 
-    for o in tape.observables:
-        is_tensor = isinstance(o, Tensor)
+    wires_list = o.wires.tolist()
+    wires = [wires_map[w] for w in wires_list]
+    if _obs_has_kernel(o):
+        return named_obs(o.name, wires)
 
-        wires = []
 
-        if is_tensor:
-            for o_ in o.obs:
-                wires_list = o_.wires.tolist()
-                w = [wires_map[w] for w in wires_list]
-                wires.append(w)
-        else:
-            wires_list = o.wires.tolist()
-            w = [wires_map[w] for w in wires_list]
-            wires.append(w)
+def _serialize_tensor_ob(ob, wires_map: dict, use_csingle: bool):
+    """Serialize a tensor observable"""
+    assert isinstance(ob, Tensor)
 
-        name = o.name if is_tensor else [o.name]
+    if use_csingle:
+        tensor_obs = TensorProdObsGPU_C64
+    else:
+        tensor_obs = TensorProdObsGPU_C128
+    return tensor_obs([_serialize_ob(o, wires_map, use_csingle) for o in ob.obs])
 
-        params = []
 
-        if not _obs_has_kernel(o):
-            if is_tensor:
-                for o_ in o.obs:
-                    if not _obs_has_kernel(o_):
-                        params.append(qml.matrix(o_).ravel().astype(ctype))
-                    else:
-                        params.append([])
-            else:
-                params.append(qml.matrix(o).ravel().astype(ctype))
+def _serialize_hamiltonian(ob, wires_map: dict, use_csingle: bool):
+    if use_csingle:
+        rtype = np.float32
+        hamiltonian_obs = HamiltonianGPU_C64
+    else:
+        rtype = np.float64
+        hamiltonian_obs = HamiltonianGPU_C128
 
-        ob = obs_py(name, params, wires)
-        obs.append(ob)
+    coeffs = np.array(ob.coeffs).astype(rtype)
+    terms = [_serialize_ob(t, wires_map, use_csingle) for t in ob.ops]
+    return hamiltonian_obs(coeffs, terms)
+
+
+def _serialize_ob(ob, wires_map, use_csingle):
+    if isinstance(ob, Tensor):
+        return _serialize_tensor_ob(ob, wires_map, use_csingle)
+    elif ob.name == "Hamiltonian":
+        return _serialize_hamiltonian(ob, wires_map, use_csingle)
+    else:
+        return _serialize_named_ob(ob, wires_map, use_csingle)
+
+
+def _serialize_observables(tape: QuantumTape, wires_map: dict, use_csingle: bool = False) -> List:
+    """Serializes the observables of an input tape.
+
+    Args:
+        tape (QuantumTape): the input quantum tape
+        wires_map (dict): a dictionary mapping input wires to the device's backend wires
+        use_csingle (bool): whether to use np.complex64 instead of np.complex128
+
+    Returns:
+        list(ObsStructC128 or ObsStructC64): A list of observable objects compatible with the C++ backend
```

Comment:
Are these return-type names correct?

Reply:
Probably this is from `pennylane-lightning` (which is probably now wrong :sweat_smile: ).

Reply:
let me check the updated one

Reply:
> Probably this is from `pennylane-lightning` (which is probably now wrong 😅 ).

Yes, it is a direct copy from `pennylane-lightning`. It seems I can't find  'ObsStructC+'